ml-agents/mlagents/trainers/trainer.py

```diff
@@ -60,7 +60,8 @@ def __init__(
         )
         self.summary_writer = tf.summary.FileWriter(self.summary_path)
         self._reward_buffer: Deque[float] = deque(maxlen=reward_buff_cap)
-        self.policy: TFPolicy = None  # type: ignore  # this will always get set
+        self.policy: TFPolicy
```

Comment:
This looks really awkward. What's the difference between self.policy and self.policies?

Reply:
self.policies is for when one trainer needs to track multiple policies (ghost trainer). But you are right, PPO/SAC never need to do this so I should bubble up this field just for multiagent/self-play algorithms.  Essentially, the get_policy and add_policy methods will access self.policies in multiagent algorithms, but ppo and sac don't ever need to. Going to fix this.

Reply:
Sounds good.

Reply:
Should we separate Trainer into SinglePolicyTrainer and MultiPolicyTrainer? Seems like they have fundamentally different interfaces.

One thing we could do is make `SinglePolicyTrainer.policy` an abstract property instead of a field. Then SACTrainer can return self.sac_policy etc.

Reply:
IMO a single policy trainer is just a special case of a multi policy trainer, with only one entry in the dict. But not sure if this PR goes that far.